File: packages/flutter_shared/lib/tools/base64_tool.dart
import 'dart:convert';
import 'dart:typed_data';

/// Base64 编解码工具
class Base64Tool {
  Base64Tool._();

  /// 文本 → Base64
  static String encode(String input) {
    if (input.isEmpty) return '';
    final bytes = utf8.encode(input);
    return base64.encode(bytes);
  }

  /// 文本 → Base64 URL Safe
  static String encodeUrlSafe(String input) {
    if (input.isEmpty) return '';
    final bytes = utf8.encode(input);
    return base64Url.encode(bytes);
  }

  /// Base64 → 文本
  static ({String result, String? error}) decode(String input) {
    if (input.isEmpty) return (result: '', error: null);
    try {
      final bytes = base64.decode(input);
      return (result: utf8.decode(bytes), error: null);
    } on FormatException catch (e) {
      return (result: '', error: '非法 Base64 字符串: ${e.message}');
    }
  }

  /// 文件字节 → Base64
  static String encodeBytes(Uint8List bytes) => base64.encode(bytes);

  /// 文件字节 → Base64 Data URI
  static String encodeDataUri(Uint8List bytes, String mimeType) {
    final b64 = base64.encode(bytes);
    return 'data:$mimeType;base64,$b64';
  }
}

File: miniapp/src/utils/tools/base64Tool.ts
/**
 * Base64 编解码工具
 * 移植自 flutter_shared/tools/base64_tool.dart
 */

export interface Base64Result {
  result: string;
  error?: string;
}

export const Base64Tool = {
  /** 文本 → Base64 */
  encode(input: string): string {
    if (!input) return '';
    return btoa(unescape(encodeURIComponent(input)));
  },

  /** 文本 → Base64 URL Safe */
  encodeUrlSafe(input: string): string {
    if (!input) return '';
    return this.encode(input)
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=+$/, '');
  },

  /** Base64 → 文本 */
  decode(input: string): Base64Result {
    if (!input) return { result: '', error: undefined };
    try {
      // 标准 base64 解码
      const decoded = atob(input);
      const result = decodeURIComponent(decoded.split('').map(c => {
        return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
      }).join(''));
      return { result, error: undefined };
    } catch (e: any) {
      // 尝试直接 ascii 解码
      try {
        const decoded = atob(input);
        return { result: decoded, error: undefined };
      } catch (e2: any) {
        return { result: '', error: `非法 Base64 字符串: ${e.message || e2.message}` };
      }
    }
  },
};
